From 4ecd97f4137d2b50ed74f3f6bf25735778b8b169 Mon Sep 17 00:00:00 2001
From: Gwenael Treuveur <gwenael.treuveur@foss.st.com>
Date: Wed, 31 Jan 2024 16:02:15 +0100
Subject: [PATCH 0793/1695] plat-stm32mp2: manage non secure reset request for
 the Cortex-M33

Don't allow NS request to manage the Cortex-M33 resets
when managed by OP-TEE

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2f23e2ef7545fcb01a492c10ed01134f4db2d79d plat-stm32mp2: manage non secure reset request for the Cortex-M33]
Signed-off-by: Gwenael Treuveur <gwenael.treuveur@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/356067
Change-Id: I1fb7aef5e736e5c423fe0c640f6fd77f924f7c97
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/375030
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gwenael TREUVEUR <gwenael.treuveur@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 .../arm/plat-stm32mp2/scmi_server_scpfw.c     | 21 +++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
index d1fe1be90..20194d3e3 100644
--- a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
+++ b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
@@ -15,6 +15,7 @@
 #include <drivers/stm32_cpu_opp.h>
 #endif
 #include <drivers/stm32_firewall.h>
+#include <drivers/stm32_remoteproc.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <initcall.h>
 #include <mm/core_memprot.h>
@@ -524,6 +525,8 @@ static TEE_Result plat_scmi_reset_assert_level(struct rstctrl *rstctrl,
 	};
 	int index = rstctrl - plat_resets;
 	paddr_t domain_iobase = 0;
+	unsigned long reset_id = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
 
 	assert(index >= 0 && (size_t)index < ARRAY_SIZE(stm32_scmi_reset));
 	domain_iobase = stm32_scmi_reset[index].base;
@@ -532,6 +535,14 @@ static TEE_Result plat_scmi_reset_assert_level(struct rstctrl *rstctrl,
 	    stm32_firewall_check_access(domain_iobase, 0, nsec_cfg))
 		return TEE_ERROR_ACCESS_DENIED;
 
+	reset_id = stm32_scmi_reset[index].reset_id;
+
+	if (reset_id == C2_R || reset_id == C2_HOLDBOOT_R) {
+		res = stm32_rproc_reset_grant_access(STM32MP2_M33_RPROC_ID);
+		if (res)
+			return res;
+	}
+
 	return rstctrl_assert_to(stm32_scmi_reset[index].rstctrl_be, to_us);
 }
 
@@ -544,6 +555,8 @@ static TEE_Result plat_scmi_reset_deassert_level(struct rstctrl *rstctrl,
 	};
 	int index = rstctrl - plat_resets;
 	paddr_t domain_iobase = 0;
+	unsigned long reset_id = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
 
 	assert(index >= 0 && (size_t)index < ARRAY_SIZE(stm32_scmi_reset));
 	domain_iobase = stm32_scmi_reset[index].base;
@@ -552,6 +565,14 @@ static TEE_Result plat_scmi_reset_deassert_level(struct rstctrl *rstctrl,
 	    stm32_firewall_check_access(domain_iobase, 0, nsec_cfg))
 		return TEE_ERROR_ACCESS_DENIED;
 
+	reset_id = stm32_scmi_reset[index].reset_id;
+
+	if (reset_id == C2_R || reset_id == C2_HOLDBOOT_R) {
+		res = stm32_rproc_reset_grant_access(STM32MP2_M33_RPROC_ID);
+		if (res)
+			return res;
+	}
+
 	return rstctrl_deassert_to(stm32_scmi_reset[index].rstctrl_be, to_us);
 }
 
-- 
2.47.2

